src/grpc/grpc_utils.h

```diff
@@ -38,6 +38,47 @@
 
 namespace triton { namespace server { namespace grpc {
 
+// The step of processing that the state is in. Every state must
+// recognize START, COMPLETE and FINISH and the others are optional.
+typedef enum {
+  // This marks the starting stage of the RPC
+  START,
+  // This marks that RPC is complete.
+  COMPLETE,
+  // This marks the stage where all the notifications from the gRPC
+  // completion queue is received and state can be safely released.
+  FINISH,
+  // This stage means that RPC has been issued to Triton for inference
+  // and is waiting for the server callbacks or cancellation to be
+  // invokded.
+  ISSUED,
+  // This stage means the request has been read from the network and
+  // can be sent to Triton for execution.
+  READ,
+  // This stage means that the response is ready to be written back to
+  // the network.
+  WRITEREADY,
```

Comment:
Can you clarify how this differs for decoupled/non-decoupled, if at all? From memory, I thought this step was more or less skipped for non-decoupled (1 response) case and it basically goes straight to WRITTEN.

Reply:
It doesn't get skipped. For non-decoupled models, the states can be in WRITEREADY and be blocked if they are not in front of the queue. For non-decoupled, we have a queue_ structure to ensure the writes are done in the order in which corresponding requests were received. 
For decoupled, we don't want requests generating large numbers of responses to block next requests that give small number of responses. Hence, we end up sending the responses as and when the response in generated by backend.